File: Session3/Java/SearchIn2DMatrix.java
import java.util.Scanner;
class SearchIn2DMatrix{  

    public static void search(int matrix[][], int target)
    {
        int n = matrix.length;
        int m = matrix[0].length;
        int s = 0;
        int e = n*m-1;
        boolean isPresent = false;
        while(s<=e) {
            int mid = (s+e)/2;
            int row = mid/m;
            int col = mid%m;
            if(matrix[row][col]==target) {
                isPresent = true;
                break;
            }
            else if(matrix[row][col] < target) {
                s = mid+1;
            }
            else {
                e = mid-1;
            }
        }
        if(isPresent==true) {
            System.out.print("True");
        }
        else {
            System.out.print("False");
        }
    }

    public static void main(String args[]){  
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        int target = sc.nextInt();
        int[][] arr = new int[n][m];
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                arr[i][j] = sc.nextInt();
            }
        }
        search(arr, target);
    }  
}  

File: Session3/C++/SearchIn2DMatrix.cpp
#include<iostream>
using namespace std;

int main(){
    
    int n,m,target;
    cin>>n>>m>>target;
    int matrix[n][m];
    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            cin>>matrix[i][j];
        }
    }

    int s = 0;
    int e = n*m-1;
    bool isPresent = false;
    while(s<=e) {
        int mid = (s+e)/2;
        int row = mid/m;
        int col = mid%m;
        if(matrix[row][col]==target) {
            isPresent = true;
            break;
        }
        else if(matrix[row][col] < target) {
            s = mid+1; // right
        }
        else {
            e = mid-1; // left
        }
    }
    if(isPresent==true) {
        cout<<"True";
    }
    else {
        cout<<"False";
    }
    return 0;
}